### PR TITLE
enable pagination for body requests

### PR DESCRIPTION
**Pull request by yinondrornahamu**
### Why?
Enabling pagination of search requests that expect parameters as body
For example:
```
GET https://service.env.globality.io/api/resource/batch
Content-Type: application/json; charset=UTF-8
{
    "identifiers": ["ID1", "ID2"... "ID1000"],
    "limit": 20,
    "offset": 0
}
```
As opposed to:
```
GET https://service.env.globality.io/api/resource?identifiers=['ID1','ID2',...'ID9000']&limit=20&offset=0
```
A typical use case is to batch requests that take a long list of identifiers as a parameter
In this case searching via URL query is not possible because of URL length limit, resulting is 414 URI too long error

### How?
Currently `all()` paging function only supports search request that expect pagination parameters in URL query (as in the second example above)
Adding a new function  `allForBodySearchRequest()` to support body search requests (as in the first example above)

### Usage Example:
```
const results = allForBodySearchRequest(req, {
        searchRequest: service.resource.search,
        body: {
            identifiers: idsList,
        },
    });
```

### CODE REVIEW

**Comment by yinondrornahamu** [comment]
Tested locally in styx by using a local copy of the function
Any idea how to test as a package?
Will add unit tests too

File: src/modules/paging/all.js [hunk 1]
```diff
@@ -1,24 +1,41 @@
...
+    const paramsValues = {
+        ...searchArgs,
+        limit,
+        offset,
+    };
+    const paramsWrapper = (sendBodyToSearchRequest) ? { body: paramsValues } : paramsValues;
+    return paramsWrapper;
+}
+
 export default async function all(
```
Comment (cdm-ium) [inline src/modules/paging/all.js]: To be honest I would probably prefer to have two separate functions, e.g. "allForSavedSearch: and normal "all"  or something (which maybe called a common function) rather than adding logic in one function to handle these cases. I realize that what you are doing is probably more common in JS but I think it's more error prone and less clear.

**Comment by yinondrornahamu** [comment]
> Key improvements:
> 
>     * add unit tests
> 
>     * consider reorganizing the code a little
> 
>     * add some comments about what this version of all is supposed to do because I think it's less obvious than our normal case
> 
> 
> Also, in my comment in your other review, I actually expected you to extend "all" for saved searches directly in the repo you were working on rather than moving to the opensource library right away. If the review feedback loop for getting this into the opensource library is too slow for you, consider adding a separate helper function directly in the project you were working in based off of `all` in nodule-openapi, rather than continuing with this pr.



> Key improvements:
> 
>     * add unit tests
> 
>     * consider reorganizing the code a little
> 
>     * add some comments about what this version of all is supposed to do because I think it's less obvious than our normal case
> 
> 
> Also, in my comment in your other review, I actually expected you to extend "all" for saved searches directly in the repo you were working on rather than moving to the opensource library right away. If the review feedback loop for getting this into the opensource library is too slow for you, consider adding a separate helper function directly in the project you were working in based off of `all` in nodule-openapi, rather than continuing with this pr.

I rather do it once in the library
After resolving the above issues how much longer should it take to get it in?

**Comment by cdm-ium** [comment]
Not long. I can push it through for you so long as we agree on the details.

**Comment by mlcohen** [comment]
> Some search endpoints get params as body (for example batch endpoints)
Extending all() pagination function to enable paginating requests in body
Currently supporting either args or body requests, not both
When it's body simply passing the pagination params in body

@yinondrornahamu it's not clear what problem you're trying to address with the above description. Please add more details. Preferably, code examples of what the current issue is plus an example of what the updated outcome will be.

WRT to locally testing your changes, please using `yalc`. See https://github.com/whitecolor/yalc

**Comment by yinondrornahamu** [comment]
> > Some search endpoints get params as body (for example batch endpoints)
> > Extending all() pagination function to enable paginating requests in body
> > Currently supporting either args or body requests, not both
> > When it's body simply passing the pagination params in body
> 
> @yinondrornahamu it's not clear what problem you're trying to address with the above description. Please add more details. Preferably, code examples of what the current issue is plus an example of what the updated outcome will be.
> 
> WRT to locally testing your changes, please using `yalc`. See https://github.com/whitecolor/yalc

I updated the description to make it more detailed and clear

**Comment by yinondrornahamu** [comment]
> Key improvements:
> 
>     * add unit tests
> 
>     * consider reorganizing the code a little
> 
>     * add some comments about what this version of all is supposed to do because I think it's less obvious than our normal case
> 
> 
> Also, in my comment in your other review, I actually expected you to extend "all" for saved searches directly in the repo you were working on rather than moving to the opensource library right away. If the review feedback loop for getting this into the opensource library is too slow for you, consider adding a separate helper function directly in the project you were working in based off of `all` in nodule-openapi, rather than continuing with this pr.

@cdm-ium I added unit tests and tested locally as a package with yalc

**Comment by mlcohen** [comment]
> Enabling pagination of search requests that expect parameters as body

A few thought:

1. Are services implemented to handle this type of query?
2. Using `GET` with a body breaks idempotency. If you're going to include a body better to use a POST with an endpoint intended to handle such a query
3. To follow up with (2), I understand what you're doing but perhaps should work within the constraints of using GET. Instead of sending all of the ID, break the IDs up into groups on the caller side and make more than one request.

**Comment by yinondrornahamu** [comment]
> > Enabling pagination of search requests that expect parameters as body
> 
> A few thoughts:
> 
>     1. Are services implemented to handle this type of query?
> 
>     2. Using `GET` with a body breaks idempotency. If you're going to include a body better to use a POST with an endpoint intended to handle such a query
> 
>     3. To follow up with (2), I understand what you're doing but perhaps should work within the constraints of using `GET`. Instead of sending all of the IDs, break the IDs up into groups on the caller side and make more than one request.

1. Yes, we have a convention in microcosm flask for this type of query, and few services using it
2. I don't know what's idempotency and how it's breaking it, but the endpoint is getting `POST` as you are suggesting
3. Considering it's using `POST` and not `GET`, I rather complete this solution that has already been implemented and tested

**Comment by mlcohen** [comment]
> I don't know what's idempotency and how it's breaking it, but the endpoint is getting POST as you are suggesting

To be idempotent is to have no additional affect when a function is called more than once with the same inputs. So when using `GET` against an endpoint, there should be no affect to the system for each call made, this compared to using `POST`, `PUT`, `DELETE` and `PATCH` which do have an affect for every call made.

In your PR's description you have the following:

```
GET https://service.env.globality.io/api/resource/batch
Content-Type: application/json; charset=UTF-8
{
    "identifiers": ["ID1", "ID2"... "ID1000"],
    "limit": 20,
    "offset": 0
}
```

That's what led me to believe you were using `GET`. Calls with get should avoided a body. So from what you said, you are in fact using `POST`, yes?

In any case, my other concern is that the logic you are adding is very specific to Globality's needs. This solution is not specific to Swagger/OpenAPI. So while, yes, this is logic we can reuse given that Globality's backend services handle it, does it really belong here?

**Comment by yinondrornahamu** [comment]
> > I don't know what's idempotency and how it's breaking it, but the endpoint is getting POST as you are suggesting
> 
> To be idempotent is to have no additional affect when a function is called more than once with the same inputs. So when using `GET` against an endpoint, there should be no affect to the system for each call made, this compared to using `POST`, `PUT`, `DELETE` and `PATCH` which do have an affect for every call made.
> 
> In your PR's description you have the following:
> 
> ```
> GET https://service.env.globality.io/api/resource/batch
> Content-Type: application/json; charset=UTF-8
> {
>     "identifiers": ["ID1", "ID2"... "ID1000"],
>     "limit": 20,
>     "offset": 0
> }
> ```
> 
> That's what led me to believe you were using `GET`. Calls with `GET` should avoid having a body. So from what you said, you are in fact using `POST`, yes?
> 
> In any case, my other concern is that the logic you are adding is very specific to Globality's needs. This solution is not specific to Swagger/OpenAPI. So while, yes, this is logic we can reuse given that Globality's backend services are designed to handle it, does it really belong here?

@mlcohen 

1. If I understand that, too keep idempotency the `GET` operation must not cause additional changes to the systems. In the above example, the `GET` returns search results based on the parameters. What are the additional changes?
2. Yes, Globality's services are using `POST` for that batch search endpoint
3. There are some examples of using `POST` for batch endpoints in the industry:
* Google Drive API - https://developers.google.com/drive/api/v3/batch
* Facebook Graph API - https://developers.facebook.com/docs/graph-api/making-multiple-requests
* Microsoft - https://docs.microsoft.com/en-us/sharepoint/dev/sp-add-ins/make-batch-requests-with-the-rest-apis?redirectedfrom=MSDN
The above endpoints are even more complex as they allow combining multiple requests into one.
Globality's current batch endpoint is a simplified version of that, targeted to batch search a particular resource. It can be extended in the future to support multiple resource requests
So in conclusion yes I think it's relevant and belongs here

**Comment by mlcohen** [comment]
> Globality's current batch endpoint is a simplified version of that, targeted to batch search a particular resource. It can be extended in the future to support multiple resource requests
So in conclusion yes I think it's relevant and belongs here

Sure, there are methods to handle batch processing. Don't disagree with that. Again, I'm not opposed to the function you've created. Rather, my hesitation is that the function you've added is: 1) Specific to how Globality manages fetching a large batch of resources; 2) the approach is not related to the OpenAPI/Swagger. Your function seems like it should exist in another library.

> In the above example, the GET returns search results based on the parameters. What are the additional changes?

Again, when using `GET` proper convention is not include a body in the request, but as you've noted, your function is using `POST`. Agree that your GET request is not causing changes to the system.

**Comment by cdm-ium** [comment]
> > Globality's current batch endpoint is a simplified version of that, targeted to batch search a particular resource. It can be extended in the future to support multiple resource requests
> > So in conclusion yes I think it's relevant and belongs here
> 
> Sure, there are methods to handle batch processing. Don't disagree with that. Again, I'm not opposed to the function you've created. Rather, my hesitation is that the function you've added is: 1) Specific to how Globality manages fetching a large batch of resources; 2) the approach is not related to the OpenAPI/Swagger. Your function seems like it should exist in another library.
> 
> > In the above example, the GET returns search results based on the parameters. What are the additional changes?
> 
> Again, when using `GET` proper convention is not include a body in the request, but as you've noted, your function is using `POST`. Agree that your GET request is not causing changes to the system.

r.e. first point, it is debatable. I know Jessie has taken the question of why we opensourced things in general. The way Yinon added it it's completely independent from the other utility funcs so it seems to fall in line with the other helpers we have in our python libs. In python these are not opensourced but they aren't really globality specific and I think that is a mistake because it makes it impossible to use these helpers (which we use EVERYWHERE in building our services) in an open source way.

for the second, it's a microcosm convention that we've been using for a while, this is just a helper function to work with that so I don't see the issue/see it as doing a new thing.

**Comment by mlcohen** [comment]
@cdm-ium Yeah, fair enough 👍🏼
